query_optimizer/ExecutionHeuristics.hpp

```diff
@@ -0,0 +1,137 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_QUERY_OPTIMIZER_EXECUTION_HEURISTICS_HPP_
+#define QUICKSTEP_QUERY_OPTIMIZER_EXECUTION_HEURISTICS_HPP_
+
+#include <vector>
+
+#include "catalog/CatalogTypedefs.hpp"
+#include "query_execution/QueryContext.pb.h"
+#include "query_optimizer/QueryPlan.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+namespace optimizer {
+
+/** \addtogroup QueryOptimizer
+ *  @{
+ */
+
+namespace execution_heuristics_internal {
+
+/**
+ * @brief A simple internal class that holds information about various
+ *        hash joins within the execution plan for a query.
+ **/
+class HashJoinInfo {
+ public:
+  HashJoinInfo(QueryPlan::DAGNodeIndex build_operator_index,
+               QueryPlan::DAGNodeIndex join_operator_index,
+               const relation_id &build_relation_id,
+               const relation_id &probe_relation_id,
+               std::vector<attribute_id> &&build_attributes,
+               std::vector<attribute_id> &&probe_attributes,
+               const QueryContext::join_hash_table_id &join_hash_table_id)
+    : build_operator_index_(build_operator_index),
+      join_operator_index_(join_operator_index),
+      build_relation_id_(build_relation_id),
+      probe_relation_id_(probe_relation_id),
+      build_attributes_(build_attributes),
+      probe_attributes_(probe_attributes),
```

Comment:
Please use the correct way for `rvalue`:

`probe_attributes_(std::move(probe_attributes)),`